In [1]:
import numpy as np
import pandas as pd
import sklearn

In [2]:
print("numpy", np.__version__)
print("pandas", pd.__version__)
print("sklearn", sklearn.__version__)

numpy 1.26.4
pandas 2.2.2
sklearn 1.4.2


## Load the datasets

In [4]:
movies_dataset = pd.read_csv("tmdb_5000_movies.csv")
credits_dataset = pd.read_csv("tmdb_5000_credits.csv")

Create copies of the datasets to analyze and proceed (good practice)

In [6]:
df_movies = movies_dataset.copy()
df_credits = credits_dataset.copy()

## Analyze the datasets

In [8]:
df_movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [9]:
df_credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [10]:
df_credits.head(1)["crew"].values

array(['[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cam

Both datasets contain information for the same movies. We can merge them into a single dataset - less hectic

In [12]:
# merged into single dataframe df based on "title" column
df = pd.merge(df_movies, df_credits, on = "title", how = "inner") # "inner" means intersection

In [13]:
df.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [14]:
df.shape

(4809, 23)

In [15]:
df["original_language"].value_counts()

original_language
en    4510
fr      70
es      32
zh      27
de      27
hi      19
ja      16
it      14
ko      12
cn      12
ru      11
pt       9
da       7
sv       5
nl       4
fa       4
th       3
he       3
ta       2
cs       2
ro       2
id       2
ar       2
vi       1
sl       1
ps       1
no       1
ky       1
hu       1
pl       1
af       1
nb       1
tr       1
is       1
xx       1
te       1
el       1
Name: count, dtype: int64

More than 95% movies are in English original language.

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

## Feature Engineering

Now, we'll remove the columns which aren't important for our analysis.

Our goal : To make content-based recommender system.

Content-based recommender creates "tags" for each movie (for recommendation). So, we'll ask ourselves that, whether a particular feature relevant in creating tag for that movie OR not.

List of important features (columns):
1) genres
2) id / movie_id (will be useful when we create website of the model & deploy it)
3) keywords
4) title
5) overview
6) cast
7) crew

In [23]:
df = df[["genres", "movie_id", "keywords", "title", "overview", "cast", "crew"]]

In [24]:
df.head()

genres  movie_id  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...     19995   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...       285   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...    206647   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...     49026   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...     49529   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

## Data pre-processing and the way ahead

We'll create a dataframe with columns - movie_id, title & tags

To create the tags column, we'll merge the columns - overview, cast, crew, keywords & genres

We'll merge the top 3 casts as we intend to recommend on the basis of top 3 casts. Same with crew, only merge director name

We notice that the features keywords, genres, cast & crew are in a weird format - we need to preprocess these also.
Any missing data in any column should also be handled.

In [30]:
# we count the missing data per column
df.isnull().sum()

genres      0
movie_id    0
keywords    0
title       0
overview    3
cast        0
crew        0
dtype: int64

In [31]:
# 3 movies have missing overview. Since, this number is very small, we shall drop these
df.dropna(inplace = True) # or you could write, df = df.dropna() directly

In [32]:
df.isnull().sum()

genres      0
movie_id    0
keywords    0
title       0
overview    0
cast        0
crew        0
dtype: int64

In [33]:
# We also check if there is any duplicate data or not
df.duplicated().sum()

0

As we can see above, zero indicates that no two rows are duplicate.

Now, we have to pre-process the columns - keywords, genres, cast & crew in "good" format

In [36]:
# viewing what the genres column looks like
df.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

genres is a list of dictionaries

required format after pre-processing is
["Action", "Adventure", "Fantasy", "Sci-Fi"]

In [39]:
# we can create a helper function to accomplish this
# def convert(obj_list): # Input is : '[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'
#     required_list = [] # will store required format after pre-processing/conversion
    
#     for dict in obj_list: # Each item in the input list is a dictionary
#         required_list.append(dict["name"])

#     return required_list

# REASON FOR COMMENTING THIS IS DUE TO A PROBLEM DESCRIBED BELOW

There's actually a problem, the input list is a STRING !!! (see '[ and ]' in the input list above)

We need to convert the string of lists into a list (so that convert() runs)

For this, we have a module in python

In [43]:
import ast
ast.literal_eval('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

See that the string of lists is converted into a single list!!!

SO, we modify our convert function like this :

In [46]:
def convert(obj_list): 
    required_list = [] 
    
    for dict in ast.literal_eval(obj_list): 
        required_list.append(dict["name"])

    return required_list

In [47]:
# Now, we need to apply this convert() to all entries in genres column
df["genres"].apply(convert) # apply() accomplishes this task

0       [Action, Adventure, Fantasy, Science Fiction]
1                        [Adventure, Fantasy, Action]
2                          [Action, Adventure, Crime]
3                    [Action, Crime, Drama, Thriller]
4                [Action, Adventure, Science Fiction]
                            ...                      
4804                        [Action, Crime, Thriller]
4805                                [Comedy, Romance]
4806               [Comedy, Drama, Romance, TV Movie]
4807                                               []
4808                                    [Documentary]
Name: genres, Length: 4806, dtype: object

In [48]:
# storing in the "genres" column
df["genres"] = df["genres"].apply(convert)

In [49]:
df.head()

genres  movie_id  \
0  [Action, Adventure, Fantasy, Science Fiction]     19995   
1                   [Adventure, Fantasy, Action]       285   
2                     [Action, Adventure, Crime]    206647   
3               [Action, Crime, Drama, Thriller]     49026   
4           [Action, Adventure, Science Fiction]     49529   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

Now, the pre-processing of "genres" is complete!!

In [51]:
# now, we'll do EXACTLY SAME pre-processing step to "keywords" column
df["keywords"] = df["keywords"].apply(convert)

In [52]:
df.head()

genres  movie_id  \
0  [Action, Adventure, Fantasy, Science Fiction]     19995   
1                   [Adventure, Fantasy, Action]       285   
2                     [Action, Adventure, Crime]    206647   
3               [Action, Crime, Drama, Thriller]     49026   
4           [Action, Adventure, Science Fiction]     49529   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

We'll pull out top 3 actors from "cast" & director from "crew"

In [54]:
df["cast"][0] # cast for the 1st movie, Avatar

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [55]:
# We want first 3 dictionaries from this (top 3 actors)
# we slightly modify the convert() function to accomplish this
def convert_cast(obj_list):
    required_list = [] 
    count = 0
    
    for dict in ast.literal_eval(obj_list): 
        if count < 3:
            required_list.append(dict["name"])
            count += 1
        else:
            break
            
    return required_list

In [56]:
# apply the transformation
df["cast"] = df["cast"].apply(convert_cast)

In [57]:
df.head()

genres  movie_id  \
0  [Action, Adventure, Fantasy, Science Fiction]     19995   
1                   [Adventure, Fantasy, Action]       285   
2                     [Action, Adventure, Crime]    206647   
3               [Action, Crime, Drama, Thriller]     49026   
4           [Action, Adventure, Science Fiction]     49529   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   
3      [Christian Bale, Michael Caine, Gary Oldman]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

Now, "crew" is left. Let's do it!!

In [59]:
df["crew"][0] # Avatar crew

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [60]:
# In this, we want that dictionary, in which job = "Director", and "name" of that dictionary is to be extracted
# we create a similar helper function

def convert_crew(obj_list):
    required_list = [] 
    
    for dict in ast.literal_eval(obj_list):
        if dict["job"] == "Director":
            required_list.append(dict["name"])
            break
            
    return required_list

In [61]:
df["crew"] = df["crew"].apply(convert_crew)

In [62]:
df.head()

genres  movie_id  \
0  [Action, Adventure, Fantasy, Science Fiction]     19995   
1                   [Adventure, Fantasy, Action]       285   
2                     [Action, Adventure, Crime]    206647   
3               [Action, Crime, Drama, Thriller]     49026   
4           [Action, Adventure, Science Fiction]     49529   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

In [63]:
df["overview"][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

Notice that the "overview" column has string. As other columns ("genres", "keywords", "cast", "crew") are lists, we convert the "overview" column also, into a list. This makes the concatenation (for creating tags) easy in the future.

In [65]:
df["overview"].apply(lambda x : x.split())

0       [In, the, 22nd, century,, a, paraplegic, Marin...
1       [Captain, Barbossa,, long, believed, to, be, d...
2       [A, cryptic, message, from, Bond’s, past, send...
3       [Following, the, death, of, District, Attorney...
4       [John, Carter, is, a, war-weary,, former, mili...
                              ...                        
4804    [El, Mariachi, just, wants, to, play, his, gui...
4805    [A, newlywed, couple's, honeymoon, is, upended...
4806    ["Signed,, Sealed,, Delivered", introduces, a,...
4807    [When, ambitious, New, York, attorney, Sam, is...
4808    [Ever, since, the, second, grade, when, he, fi...
Name: overview, Length: 4806, dtype: object

In [66]:
df["overview"] = df["overview"].apply(lambda x : x.split())

In [67]:
df.head()

genres  movie_id  \
0  [Action, Adventure, Fantasy, Science Fiction]     19995   
1                   [Adventure, Fantasy, Action]       285   
2                     [Action, Adventure, Crime]    206647   
3               [Action, Crime, Drama, Thriller]     49026   
4           [Action, Adventure, Science Fiction]     49529   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

In [68]:
# Transformation to be applied on 4 columns, "genres", "keywords", "cast" & "crew"
# Sam Worthington
# SamWorthington -> required tag so that the recommender system searches for SamWorthington movies, instead of just Sam

In [69]:
df["genres"].apply(lambda x : [i.replace(" ", "") for i in x])

0       [Action, Adventure, Fantasy, ScienceFiction]
1                       [Adventure, Fantasy, Action]
2                         [Action, Adventure, Crime]
3                   [Action, Crime, Drama, Thriller]
4                [Action, Adventure, ScienceFiction]
                            ...                     
4804                       [Action, Crime, Thriller]
4805                               [Comedy, Romance]
4806               [Comedy, Drama, Romance, TVMovie]
4807                                              []
4808                                   [Documentary]
Name: genres, Length: 4806, dtype: object

In [70]:
df["genres"] = df["genres"].apply(lambda x : [i.replace(" ", "") for i in x])
df["keywords"] = df["keywords"].apply(lambda x : [i.replace(" ", "") for i in x])
df["cast"] = df["cast"].apply(lambda x : [i.replace(" ", "") for i in x])
df["crew"] = df["crew"].apply(lambda x : [i.replace(" ", "") for i in x])

In [71]:
df.head()

genres  movie_id  \
0  [Action, Adventure, Fantasy, ScienceFiction]     19995   
1                  [Adventure, Fantasy, Action]       285   
2                    [Action, Adventure, Crime]    206647   
3              [Action, Crime, Drama, Thriller]     49026   
4           [Action, Adventure, ScienceFiction]     49529   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                            cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]  
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]  
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]

Now, creating a "tags" column, which is a concatenation of above 5 columns

In [73]:
# concatenation using the '+' operator
df["tags"] = df["genres"] + df["keywords"] + df["overview"] + df["cast"] + df["crew"]

In [74]:
df.head()

genres  movie_id  \
0  [Action, Adventure, Fantasy, ScienceFiction]     19995   
1                  [Adventure, Fantasy, Action]       285   
2                    [Action, Adventure, Crime]    206647   
3              [Action, Crime, Drama, Thriller]     49026   
4           [Action, Adventure, ScienceFiction]     49529   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

                                                tags  
0  [Action, Adventure, Fantasy, ScienceFiction, c...  
1  [Adventure, Fantasy, Action, ocean, drugabuse,...  
2  [Action, Adventure, Crime, spy, basedonnovel, ...  
3  [Action, Crime, Drama, Thriller, dccomics, cri...  
4  [Action, Adventure, ScienceFiction, basedonnov...

In [75]:
df["tags"][0]

['Action',
 'Adventure',
 'Fantasy',
 'ScienceFiction',
 'cultureclash',
 'future',
 'spacewar',
 'spacecolony',
 'society',
 'spacetravel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alienplanet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'loveaffair',
 'antiwar',
 'powerrelations',
 'mindandsoul',
 '3d',
 'In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.',
 'SamWorthington',
 'ZoeSaldana',
 'SigourneyWeaver',
 'JamesCameron']

The above list of strings are the tags for the 1st movie - Avatar!!!

Now, we don't need the 5 columns from which "tags" was made. So, let's remove them.

In [78]:
df = df.drop(["genres", "keywords", "overview", "cast", "crew"], axis = 1)

In [79]:
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [Action, Adventure, Fantasy, ScienceFiction, c...  
1  [Adventure, Fantasy, Action, ocean, drugabuse,...  
2  [Action, Adventure, Crime, spy, basedonnovel, ...  
3  [Action, Crime, Drama, Thriller, dccomics, cri...  
4  [Action, Adventure, ScienceFiction, basedonnov...

Now, we can also convert the list present in "tags" to a string.

In [81]:
df["tags"] = df["tags"].apply(lambda x : " ".join(x))

In [82]:
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  Action Adventure Fantasy ScienceFiction cultur...  
1  Adventure Fantasy Action ocean drugabuse exoti...  
2  Action Adventure Crime spy basedonnovel secret...  
3  Action Crime Drama Thriller dccomics crimefigh...  
4  Action Adventure ScienceFiction basedonnovel m...

In [83]:
df["tags"][0]

'Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [84]:
df["tags"] = df["tags"].apply(lambda x : x.lower())

In [85]:
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  action adventure fantasy sciencefiction cultur...  
1  adventure fantasy action ocean drugabuse exoti...  
2  action adventure crime spy basedonnovel secret...  
3  action crime drama thriller dccomics crimefigh...  
4  action adventure sciencefiction basedonnovel m...

At this point, we've the tags for each movie.

## The Way Ahead

Our goal is to make a website in which a user will tell a movie's name and we have to give him/her a list of 5 similar movies.

Similarity is to be measured in terms of tags.

In [90]:
df["tags"][0]

'action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. samworthington zoesaldana sigourneyweaver jamescameron'

In [91]:
df["tags"][1]

"adventure fantasy action ocean drugabuse exoticisland eastindiatradingcompany loveofone'slife traitor shipwreck strongwoman ship alliance calypso afterlife fighter pirate swashbuckler aftercreditsstinger captain barbossa, long believed to be dead, has come back to life and is headed to the edge of the earth with will turner and elizabeth swann. but nothing is quite as it seems. johnnydepp orlandobloom keiraknightley goreverbinski"

To check if the above 2 movies are similar or not, we need to calculate the similarity score of above 2 texts (tags).

One way to calculate similarity score of above 2 texts is to find the number of common words. If this number is greater than a threshold, then we can say that the 2 movies are similar to each other. BUT, this way is not that good.

Here comes the use of text vectorization.

## Text Vectorization

Text vectorization is the process of converting textual data into numerical vectors that can be used as input for machine learning models. Since machine learning algorithms work with numerical data, text must be transformed into a mathematical representation to analyze, classify, or extract insights from it.

When we convert text to vectors, there would be almost 5000 vectors (representing each movie) in 2D space. When a user tells that he/she likes a particular movie, then we would recommend him/her to watch that movies (vectors) which are CLOSEST to the input movie vector. This is why we'll use text vectorization here.

There are many techniques to perform text vectorization :
1) Bag of Words (BoW)
2) TF-IDF
and many more....

We'll use the simplest and the most famous technique of BoW here. (https://youtu.be/1xtrIEwY_zY?feature=shared&t=3600)

In [100]:
## Stemming (why ? is described below)
import nltk

In [101]:
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [102]:
def stem(text):
    y = []

    for i in text.split():
        y.append(ps.stem(i))

    return  " ".join(y)

In [103]:
df["tags"] = df["tags"].apply(stem)

In [104]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 5000, stop_words = "english")

In [105]:
movie_vectors = cv.fit_transform(df["tags"]).toarray()

In [106]:
movie_vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [107]:
# this is the first movie vector
movie_vectors[0] 

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [108]:
# we can see the 5000 most frequent words 
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zone', 'zoo', 'zooeydeschanel'],
      dtype=object)

Main problem is that in the 5000 most frequent words, we get the same words again & again. For example, both "action" & "actions" appear. Also, "actor" & "actors" appear. To avoid this, we did stemming.

In [110]:
print(ps.stem("loving"))
print(ps.stem("loved"))
print(ps.stem("love"))

love
love
love


In [111]:
print(ps.stem("dancing"))
print(ps.stem("dance"))
print(ps.stem("dancer"))

danc
danc
dancer


We now have 4806 vectors, each with 5000 components. We now have to calculate the distance of one movie with every other movie. Larger the distance, more dissimliar are the movies.

Note that we wouldn't calculate the euclidean distance between any two movies. Instead, we calculate the cosine distance between them (i.e. the angle between the two movie vectors).

In higher dimensionality, (here, 5000 dimensional), euclidean distance often fails.

In [115]:
from sklearn.metrics.pairwise import cosine_similarity

In [116]:
similarity = cosine_similarity(movie_vectors)

In [117]:
similarity.shape

(4806, 4806)

In [118]:
# similarity[0] will tell the similarity of first movie with every other movie
similarity[0]

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ])

Logically, first movie should have a similarity score of 1 with itself. That's why the first element of above numpy array is 1

So, the similarity matrix has all its diagonal entries = 1.

In [121]:
def recommend(movie_name):
    movie_index = df[df["title"] == movie_name].index[0]
    distances = similarity[movie_index]
    recommended_movies_list = sorted(list(enumerate(distances)), reverse = True, key = lambda x : x[1])[1 : 6]

    for recommended_movie in recommended_movies_list:
        print(df.iloc[recommended_movie[0]].title)

In [122]:
recommend("Avatar")

Aliens vs Predator: Requiem
Aliens
Falcon Rising
Independence Day
Titan A.E.


In [123]:
recommend("Batman Begins")

The Dark Knight
Batman
Batman
The Dark Knight Rises
10th & Wolf


In [124]:
recommend("Batman")

Batman
Batman & Robin
Batman Begins
Batman Returns
The R.M.


In [125]:
import pickle

In [126]:
pickle.dump(df, open("movies.pkl", "wb"))

In [127]:
df["title"].values

array(['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre',
       ..., 'Signed, Sealed, Delivered', 'Shanghai Calling',
       'My Date with Drew'], dtype=object)

In [128]:
pickle.dump(similarity, open("similarity.pkl", "wb"))